Propose a new ADR to use React, superseding ADR001. (#59)
mdn/mdn: ADRs/004-use-react.md

## Use React for the MDN Frontend

|Status        | proposed    |
|--------------|-------------|
|**Proposed**  | 2019-02-13
|**Accepted**  | 
|**Supersedes**| ADR 001
|**Driver**    | David Flanagan
|**Approver**  | John Whitlock
|**Consulted** | Schalk Neethling, Justin Fagnani
|**Informed**  | MDN core team and contributors

### Decision

We will use the React framework and related technologies when
developing significant new frontend features for the MDN website.
We may also convert existing features to React as time and
opportunity allows, but this ADR does not mandate that.

### Context

The Kuma frontend is currently implemented using jQuery and, to a
lesser extent, jQueryUI, but without a modern frontend framework.  As
a result our UX (the menus in the navigation header, for example)
feels dated. The development team feels that our development velocity
is hampered by lack of a framework. Furthermore, we believe that our
ability to collaborate with contractors and with open-source
contributors would be enhanced if we were using a modern framework
with a large developer community.

> This ADR supersedes ADR-001. It arrives at the same decision, but

Author: This paragraph is new.

- [testpilot.firefox.com](https://testpilot.firefox.com/)
  ([mozilla/testpilot master](https://github.com/mozilla/testpilot/tree/master),
  before [end-of-life](https://github.com/mozilla/testpilot/tree/eol))
- [donate.mozilla.org](https://donate.mozilla.org)
  ([mozilla/donate.mozilla.org](https://github.com/mozilla/donate.mozilla.org))
- [hubs.mozilla.com](https://hubs.mozilla.com/)
  ([mozilla/hubs](https://github.com/mozilla/hubs))
- [data.firefox.com](https://data.firefox.com/)
  ([mozilla/ensemble](https://github.com/mozilla/ensemble))
- [learning.mozilla.org](https://learning.mozilla.org)
  ([mozilla/learning.mozilla.org](https://github.com/mozilla/learning.mozilla.org))
- [changecopyright.org](https://changecopyright.org)
  ([mozilla/copyright](https://github.com/mozilla/copyright))
- [foundation.mozilla.org](https://foundation.mozilla.org)
  ([mozilla/foundation.mozilla.org](https://github.com/mozilla/foundation.mozilla.org))
- [webmaker.org](https://webmaker.org)
  ([mozilla/webmaker.org](https://github.com/mozilla/webmaker.org))

I was not able to find any Mozilla websites that use Angular, Vue or Polymer.

> #### Web Components

Author: I've moved this section up from the bottom of the file, and have rewritten the three paragraphs below. Hopefully there is no misinformation here anymore.

> "Web Components" is the collective name for a set of technologies that

Reviewer: ```suggestion
"Web Components" is the collective name for a set of standards that
```

"standards" because it's an important difference vs various user-land technologies.

Author: I did the other edit instead.

facilitate the creation of reusable UI components. Web Components are

> based on web standards, and in that sense they seem like the best

Reviewer: ```suggestion
are web standards, and in that sense they seem like the best
```

Maybe this edit, since the name refers to the actual standards? 🤷‍♂️

Author: I made this edit. Thanks.

choice for Mozilla.

> On the other hand, browser support is currently incomplete and

Reviewer: I don't quite want to get in a back and forth on phrasing here, since it's your project and document, so obviously feel free to disregard, but:

While "incomplete" is technically correct, browser support is quite good right now with Chrome, Safari, Firefox, and Edge's new preview. We're at >90% mobile and >80% desktop according to caniuse.com. I would not be surprised if MDNs numbers were even better.

Additionally, the polyfills work back to IE11, Edge, Safari 9 and Chrome 41.

Author: I tweaked this slightly.

polyfills are still required. Also, Web Components do not yet have the

> tools, ecosystem, or developer community that React has.

Reviewer: I'm not sure of any tools that don't support web components - they're just JS libraries. Ecosystem and developer community I agree with.

Author: I didn't have specific tools in mind: was thinking about things like the large world of plugins for eslint and webpack. With React, if you can think of a tool you want, it is probably on NPM. I'm assuming that that isn't quite as true for Web Components. But since that is just an assumption, I'll just cut the mention of tools.

- [addons.mozilla.org](https://addons.mozilla.org/)
  ([mozilla/addons-frontend](https://github.com/mozilla/addons-frontend/))
- [testpilot.firefox.com](https://testpilot.firefox.com/)
  ([mozilla/testpilot master](https://github.com/mozilla/testpilot/tree/master),
  before [end-of-life](https://github.com/mozilla/testpilot/tree/eol))
- [donate.mozilla.org](https://donate.mozilla.org)
  ([mozilla/donate.mozilla.org](https://github.com/mozilla/donate.mozilla.org))
- [hubs.mozilla.com](https://hubs.mozilla.com/)
  ([mozilla/hubs](https://github.com/mozilla/hubs))
- [data.firefox.com](https://data.firefox.com/)
  ([mozilla/ensemble](https://github.com/mozilla/ensemble))
- [learning.mozilla.org](https://learning.mozilla.org)
  ([mozilla/learning.mozilla.org](https://github.com/mozilla/learning.mozilla.org))
- [changecopyright.org](https://changecopyright.org)
  ([mozilla/copyright](https://github.com/mozilla/copyright))
- [foundation.mozilla.org](https://foundation.mozilla.org)
  ([mozilla/foundation.mozilla.org](https://github.com/mozilla/foundation.mozilla.org))
- [webmaker.org](https://webmaker.org)
  ([mozilla/webmaker.org](https://github.com/mozilla/webmaker.org))

> I was not able to find any Mozilla websites that use Angular, Vue or Polymer.

Reviewer: Firefox just launched support publically with 63, so as far as websites go, this shouldn't be surprising.

However, does the browser itself count? The entire Firefox UI is being rewritten in web components. I believe Firefox is shipping web component in the UI in production already (since at least 62 I think).

* The design review packet: https://mozilla.github.io/firefox-browser-architecture/text/0007-xbl-design-review-packet.html
* The meta-bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1397874
* "Are We XBL Still?" is a site that visualizes the progress: https://bgrins.github.io/xbl-analysis/

> I was not able to find any Mozilla websites that use Angular, Vue or Polymer.

Reviewer 2: I read https://mozilla.github.io/firefox-browser-architecture/text/0007-xbl-design-review-packet.html differently than you do. Maybe the strategy changed since it was written. It looks like Web Components and Custom Elements are used for some items, and "unrefactored" bits for others. Larger components, like [debugger.html](https://github.com/devtools-html/debugger.html), were re-written in React.

The other documents like "Are We XBL Still?" are focused on the XBL deprecation, and don't seem to add much. Is [Are We XBL Still?](https://bgrins.github.io/xbl-analysis/) an example of a site implemented with Web Components? It looks like ``<div>`` soup to me, and poor commit comments. It's meant for display and rapid updating, not for reading the code.

Maybe the sentence should be moved to each section. For example, in the Angular section, say "I was not able to find any Mozilla websites that use Angular", rather than collect these all under React.

Then, for Web Components, say "Firefox is in the process of [removing XBL](https://mozilla.github.io/firefox-browser-architecture/text/0007-xbl-design-review-packet.html) (eXtensible Bindings Language), used to define user interfaces elements in the browser. Web Components, specifically Custom Elements, have replaced many XBL bindings."

Or, stay silent about it.

> I was not able to find any Mozilla websites that use Angular, Vue or Polymer.

Reviewer: In the design packet the entire "Web Components" section talks about converting XUL components to web components. That's happening now.

"Are We XBL Still?" itself isn't an example of web components usage, it's just documenting the web components usage in the Firefox UI. The bindings tagged with #custom-element have been converted to web components, like: https://bugzilla.mozilla.org/show_bug.cgi?id=1527105

> I was not able to find any Mozilla websites that use Angular, Vue or Polymer.

Author: I'm not going to mention the XBL to WC thing because that is, by definition, Gecko only, and not proof of suitability for use on websites. (For the same reason i'm not mention the web components work we did long ago on FirefoxOS).

I moved the "could not find" bit about Angular and Vue to their subsections, and just dropped the Polymer mention.

#### Web Components

"Web Components" is the collective name for a set of technologies that
facilitate the creation of reusable UI components. Web Components are
web standards, and in that sense they seem like the best choice for
Mozilla.

On the other hand, browser support is not quite complete yet and
polyfills are still required for older browsers. Also, Web Components
do not yet have the large ecosystem and developer community that React
has.

> If we didn't have such ambitous goals for 2019 and weren't expecting

Author: done

to work with contractors on those goals, I would want to seriously
consider using Web Components because they seem like the best solution

> for the open web. But for pragmatic reasons related 2019 goals and

Author: Took out the But and added a "however" later.

- [learning.mozilla.org](https://learning.mozilla.org)
  ([mozilla/learning.mozilla.org](https://github.com/mozilla/learning.mozilla.org))
- [changecopyright.org](https://changecopyright.org)
  ([mozilla/copyright](https://github.com/mozilla/copyright))
- [foundation.mozilla.org](https://foundation.mozilla.org)
  ([mozilla/foundation.mozilla.org](https://github.com/mozilla/foundation.mozilla.org))
- [webmaker.org](https://webmaker.org)
  ([mozilla/webmaker.org](https://github.com/mozilla/webmaker.org))

#### Web Components

"Web Components" is the collective name for a set of technologies that
facilitate the creation of reusable UI components. Web Components are
web standards, and in that sense they seem like the best choice for
Mozilla.

On the other hand, browser support is not quite complete yet and
polyfills are still required for older browsers. Also, Web Components
do not yet have the large ecosystem and developer community that React
has.

If we didn't have such ambitious goals for 2019 and weren't expecting
to work with contractors on those goals, I would want to seriously
consider using Web Components because they seem like the best solution

> for the open web. For pragmatic reasons related 2019 goals and

Reviewer 2: "For pragmatic reasons related 2019 goals and staffing" still doesn't sound right when I read it aloud. "For pragmatic reasons related **to** 2019 goals and staffing" would sound better, but also seems to repeat the first sentence of the paragraph. Can you revise again @davidflanagan?

Author: Fixed, I hope. Sorry it took me so long to respond to this.
